### DISCUSSION TITLE
'fuzzy' Scala version switch

### DISCUSSION THREAD

**Discussion by raboof** [sbt/sbt#6893] [Ideas]
In projects with `crossScalaVersions`, you can use the ‘switch’ command (e.g. `sbt ++2.13.7 test`) to test the modules that support that Scala version.

It would be neat to have a more ‘fuzzy’ variation of this command so you could do `sbt "+~ 2.13 test"` or `sbt +~ 3 test`, which would pick the first entry from `crossScalaVersions` that matches that prefix (if any) and builds with that. That way there’s less chance of CI scripts or command-line usage to become inconsistent with `crossScalaVersions`.

(we're trying this out in https://github.com/akka/akka-grpc/blob/main/project/Common.scala#L82)

**Comment by eed3si9n** [comment]
Good idea. I don't want to increase the number of symbols if we can help it, so maybe we should treat the argument as glob expression and accept `2.13.*` and `3.*` as argument and fail if `2.*` returns more than one Scala versions, or something.

**Comment by raboof** [comment]
sketched it out a bit in https://github.com/sbt/sbt/pull/6894

**Comment by Nirvikalpa108** [comment]
Thank you so much for this great contribution @raboof 🎉.  Just out of interest, what was your thinking around [producing an error](https://github.com/sbt/sbt/pull/6894/files#diff-d3593a6d032d1cfb50286b5f01c257c9c740200a1e0862d6482196968be3f7f7R336-R338) when there are multiple matches in the `crossScalaVersions` list?

If I want to test my sbt project with `2.13.*` and I have a `crossScalaVersions` list of `Seq("2.13.15", "2.13.16")`, although perhaps an edge case, I would like one to be chosen for me instead of sbt producing an error. What are your thoughts?

**Reply by raboof** [reply]
> Thank you so much for this great contribution @raboof :tada:. 

Glad you like it! :)

> Just out of interest, what was your thinking around [producing an error](https://github.com/sbt/sbt/pull/6894/files#diff-d3593a6d032d1cfb50286b5f01c257c9c740200a1e0862d6482196968be3f7f7R336-R338) when there are multiple matches in the `crossScalaVersions` list?
> 
> If I want to test my sbt project with `2.13.*` and I have a `crossScalaVersions` list of `Seq("2.13.15", "2.13.16")`, although perhaps an edge case, I would like one to be chosen for me instead of sbt producing an error. What are your thoughts?

That was mostly a practical reason: it wasn't obvious how to choose between multiple matching versions, and this way I didn't need to choose :). I would be OK with any choice, as long as it's deterministic. "Parsing" the versions might be tricky, but perhaps simply using the 'last match in the list' would make sense: people might list versions in ascending order and would be OK with picking the latest.

**Reply by eed3si9n** [reply]
If someone writes `++2.* test` why not test all 2.*?

**Reply by Nirvikalpa108** [reply]
Ah I completely understand @raboof, that makes sense 👍🏽

@eed3si9n would you mind explaining why you suggested this please? Is it for ease of implementation or because you think this is what users would expect? Also, do you mean that each matching version would be tested one after the other? I just want to check I understand your idea, thank you 😊

**Reply by eed3si9n** [reply]
Under some situation, it's totally conceivable to list binary *compatible* Scala versions in `crossScalaVersions`:
- Scala 3 (see also https://github.com/sbt/sbt/issues/6915) keeps only forward compatibility, so some people are cross building among 3.x
- Compiler plugins need to be published for each Scala version releases

In these situations returning multiple Scala versions for `2.13.*`, `2.*`, `3.*` makes sense for the purpose of usability in my opinion.

Failing the build because the glob pattern matches multiple version seems somewhat arbitrary since `++ test` would otherwise able to handle multiple Scala versions. (Sorry I didn't catch this during review)

**Reply by eed3si9n** [reply]
I was thinking about `+ test`, not `++`. While `++2.* test` would still be useful, implementing it might be a bit more complicated.

One strategy might be to list out all the matching versions numbers, and emit `++<sv> test` where `<sv>` is concrete version number. That's more or less how `+` is implemented based on `++`.

**Reply by raboof** [reply]
I can see how that could be useful, but perhaps that should then be  `+ 2.* test` rather than `++ 2.* test`?

**Reply by sideeffffect** [reply]
> If someone writes `++2.* test` why not test all 2.*?

And if there are no 2.* projects, not tests should be run, right?

**Reply by eed3si9n** [reply]
`++` is fundamentally a switch command, so if it's doesn't match any subprojects it should always error.

**Reply by sideeffffect** [reply]
> should always error

I thought that `++_!` (note the bang at the end of the line) is for that. I'm asking about this, because filtering-like behavior for e.g. `++2.12.*` would solve an important issue I (and many other people) face. The issue being how to split the build of a big sbt project in CI, and splitting by Scala version seems like a good choice and that's where filtering behavior `++2.12.*` comes enters the picture.

Also, running no task if `++2.12.*` doesn't match anything would be consistent with running multiple if there are multiple matches.
/cc @armanbilge @keynmol

**Reply by armanbilge** [reply]
@sideeffffect I agree that querying/filtering features would be very beneficial, but I don't think `++` should do that. That idea is discussed in https://github.com/sbt/sbt/discussions/6801.

> The issue being how to split the build of a big sbt project in CI

@sideeffffect I recommend you look at how we do it for [sbt-typelevel](https://typelevel.org/sbt-typelevel/). It automatically splits CI by Scala version and platform (JVM, JS, Native).

**Comment by eed3si9n** [comment]
Here's another improvement idea - https://github.com/sbt/sbt/issues/6934
